File: mobile/ios/FrankenRemoteApp/Sources/Lifecycle/AppLifecycleManager.swift
//
// AppLifecycleManager.swift — Mobile Lifecycle & Authority Management
//
// Enforces Constitutional Invariants from Plan §16.2:
// 1. Backgrounding immediately releases remote input control and invalidates leases.
// 2. Resuming obtains a fresh lease and requests a recovery IDR frame.
// 3. Network transitions (Wi-Fi/Cellular) trigger route re-evaluation; never preserve stale leases.
// 4. Input queues and IME compositions are cancelled on backgrounding; never replayed.
//

import Foundation
import SwiftUI
import Network
import Combine

@MainActor
public final class AppLifecycleManager: ObservableObject {
    @Published public private(set) var isNetworkAvailable: Bool = true
    @Published public private(set) var isCellular: Bool = false
    @Published public private(set) var currentScenePhase: ScenePhase = .active

    private let pathMonitor = NWPathMonitor()
    private let monitorQueue = DispatchQueue(label: "com.frankenremote.network_monitor")

    public var onBackgroundEntered: (() -> Void)?
    public var onForegroundResumed: (() -> Void)?
    public var onNetworkRouteChanged: (() -> Void)?

    public init() {
        startNetworkMonitoring()
    }

    deinit {
        pathMonitor.cancel()
    }

    public func handleScenePhaseChange(_ newPhase: ScenePhase) {
        let oldPhase = currentScenePhase
        currentScenePhase = newPhase

        switch (oldPhase, newPhase) {
        case (_, .background):
            // Plan §16.2: Backgrounding client immediately releases control and drops queued input.
            onBackgroundEntered?()

        case (.background, .active):
            // Plan §16.2: Resuming obtains a fresh lease and recovery frame; never replays old inputs.
            onForegroundResumed?()

        default:
            break
        }
    }

    private func startNetworkMonitoring() {
        pathMonitor.pathUpdateHandler = { [weak self] path in
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                let available = path.status == .satisfied
                let cellular = path.isExpensive || path.usesInterfaceType(.cellular)

                let routeChanged = (self.isCellular != cellular) || (self.isNetworkAvailable != available)

                self.isNetworkAvailable = available
                self.isCellular = cellular

                if routeChanged {
                    // Plan §16.2: Network transitions re-evaluate the route without preserving stale leases.
                    self.onNetworkRouteChanged?()
                }
            }
        }
        pathMonitor.start(queue: monitorQueue)
    }
}

File: mobile/android/app/src/main/kotlin/com/frankenremote/app/lifecycle/AppLifecycleManager.kt
package com.frankenremote.app.lifecycle

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AppLifecycleManager(context: Context) : DefaultLifecycleObserver {

    private val connectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    private val _isNetworkAvailable = MutableStateFlow(true)
    val isNetworkAvailable: StateFlow<Boolean> = _isNetworkAvailable.asStateFlow()

    private val _isCellular = MutableStateFlow(false)
    val isCellular: StateFlow<Boolean> = _isCellular.asStateFlow()

    var onBackgroundEntered: (() -> Unit)? = null
    var onForegroundResumed: (() -> Unit)? = null
    var onNetworkRouteChanged: (() -> Unit)? = null

    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            val caps = connectivityManager.getNetworkCapabilities(network)
            val isCell = caps?.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) == true
            _isNetworkAvailable.value = true
            _isCellular.value = isCell
            onNetworkRouteChanged?.invoke()
        }

        override fun onLost(network: Network) {
            _isNetworkAvailable.value = false
            onNetworkRouteChanged?.invoke()
        }
    }

    init {
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        connectivityManager.registerNetworkCallback(request, networkCallback)
    }

    fun unregister() {
        try {
            connectivityManager.unregisterNetworkCallback(networkCallback)
        } catch (_: Exception) {}
    }

    override fun onStop(owner: LifecycleOwner) {
        // Plan §16.2: Backgrounding client immediately releases control and drops queued input.
        onBackgroundEntered?.invoke()
    }

    override fun onResume(owner: LifecycleOwner) {
        // Plan §16.2: Resuming obtains a fresh lease and recovery frame; never replays old inputs.
        onForegroundResumed?.invoke()
    }
}
